File: src/roma/gvisor/udf/sample_udf.cc
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <cmath>
#include <cstring>
#include <iostream>
#include <vector>

#include "absl/log/check.h"
#include "absl/log/initialize.h"
#include "absl/log/log.h"
#include "absl/strings/numbers.h"
#include "google/protobuf/util/delimited_message_util.h"
#include "src/roma/gvisor/host/callback.pb.h"
#include "src/roma/gvisor/udf/sample.pb.h"

using ::privacy_sandbox::server_common::gvisor::Callback;
using ::privacy_sandbox::server_common::gvisor::FUNCTION_CALLBACK;
using ::privacy_sandbox::server_common::gvisor::FUNCTION_HELLO_WORLD;
using ::privacy_sandbox::server_common::gvisor::FUNCTION_PRIME_SIEVE;
using ::privacy_sandbox::server_common::gvisor::
    FUNCTION_TEN_CALLBACK_INVOCATIONS;
using ::privacy_sandbox::server_common::gvisor::SampleRequest;
using ::privacy_sandbox::server_common::gvisor::SampleResponse;

// Find all prime numbers less than this:
constexpr int kPrimeCount = 100'000;

void RunHelloWorld(SampleResponse& bin_response) {
  bin_response.set_greeting("Hello, world!");
}

void RunPrimeSieve(SampleResponse& bin_response) {
  // Create a boolean array of size n+1
  std::vector<bool> primes(kPrimeCount + 1, true);
  // Set first two values to false
  primes[0] = false;
  primes[1] = false;
  // Loop through the elements
  for (int i = 2; i <= sqrt(kPrimeCount); i++) {
    if (primes[i]) {
      for (int j = i * i; j <= kPrimeCount; j += i) {
        primes[j] = false;
      }
    }
  }
  // Loop through the array from 2 to n
  for (int i = 2; i <= kPrimeCount; i++) {
    if (primes[i]) {
      bin_response.add_prime_number(i);
    }
  }
}

void RunEchoCallback(int comms_fd) {
  Callback callback;
  callback.set_function_name("example");
  CHECK(google::protobuf::util::SerializeDelimitedToFileDescriptor(callback,
                                                                   comms_fd));
  google::protobuf::io::FileInputStream input(comms_fd);
  CHECK(google::protobuf::util::ParseDelimitedFromZeroCopyStream(
      &callback, &input, nullptr));
}

int main(int argc, char* argv[]) {
  absl::InitializeLog();
  if (argc < 3) {
    std::cerr << "Not enough arguments!";
    return -1;
  }
  int32_t write_fd;
  CHECK(absl::SimpleAtoi(argv[1], &write_fd))
      << "Conversion of write file descriptor string to int failed";
  int comms_fd;
  CHECK(absl::SimpleAtoi(argv[2], &comms_fd))
      << "Conversion of comms file descriptor string to int failed";

  SampleRequest bin_request;
  bin_request.ParseFromIstream(&std::cin);
  SampleResponse bin_response;
  switch (bin_request.function()) {
    case FUNCTION_HELLO_WORLD:
      RunHelloWorld(bin_response);
      break;
    case FUNCTION_PRIME_SIEVE:
      RunPrimeSieve(bin_response);
      break;
    case FUNCTION_CALLBACK:
      RunEchoCallback(comms_fd);
      break;
    case FUNCTION_TEN_CALLBACK_INVOCATIONS:
      for (int i = 0; i < 10; ++i) {
        RunEchoCallback(comms_fd);
      }
      break;
    default:
      LOG(INFO) << "Invalid function enum.";
  }
  PCHECK(::close(comms_fd) == 0);

  bin_response.SerializeToFileDescriptor(write_fd);
  return 0;
}

File: src/roma/gvisor/udf/sample_udf.go
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bufio"
	proto "github.com/golang/protobuf/proto"
	pb "github.com/privacysandbox/data-plane-shared/apis/roma/binary/example"
	"log"
	"math"
	"os"
	"strconv"
)

const kPrimeCount = 100_000

func runHelloWorld(binResponse *pb.GetValuesResponse) {
	message := "Hello, world from Go!"
	binResponse.Greeting = []byte(message)
}

func runPrimeSieve(binResponse *pb.GetValuesResponse) {
	// Create a boolean slice of size kPrimeCount+1
	primes := make([]bool, kPrimeCount+1)
	for i := range primes {
		primes[i] = true
	}

	primes[0] = false
	primes[1] = false

	for i := 2; i <= int(math.Sqrt(float64(kPrimeCount))); i++ {
		if primes[i] {
			for j := i * i; j <= kPrimeCount; j += i {
				primes[j] = false
			}
		}
	}

	for i := 2; i <= kPrimeCount; i++ {
		if primes[i] {
			binResponse.PrimeNumber = append(binResponse.PrimeNumber, int32(i))
		}
	}
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Not enough input arguments")
	}
	writeFd, err := strconv.Atoi(os.Args[1])
	if err != nil {
		log.Fatal("Missing write file descriptor: ", err)
	}

	binRequest := &pb.GetValuesRequest{}
	buf := make([]byte, 10)
	reader := bufio.NewReader(os.Stdin)
	n, err := reader.Read(buf)
	if err != nil {
		log.Fatal("Failed to read from stdin: ", err)
	}
	if err := proto.Unmarshal(buf[:n], binRequest); err != nil {
		log.Fatal("Failed to parse GetValuesRequest: ", err)
	}

	binResponse := &pb.GetValuesResponse{}
	switch binRequest.Function {
	case pb.FunctionType_FUNCTION_HELLO_WORLD:
		runHelloWorld(binResponse)
	case pb.FunctionType_FUNCTION_PRIME_SIEVE:
		runPrimeSieve(binResponse)
	default:
		log.Print("Unexpected input")
		return
	}

	writeFile := os.NewFile(uintptr(writeFd), "pipe")
	defer writeFile.Close()
	output, err := proto.Marshal(binResponse)
	if err != nil {
		log.Fatal("Failed to serialize GetValuesResponse: ", err)
	}
	_, err = writeFile.Write(output)
	if err != nil {
		log.Print("Failed to write output to pipe: ", err)
	}
}
